Allow longer CIDs (#2749)
quicwg/base-drafts: draft-ietf-quic-transport.md

tokens that are sent in Retry packets as they are carried in the same field.

The token MUST NOT include information that would allow it to be linked by an

> on-path observer to the connection on which it was issued. For example, it

Reviewer: ```suggestion
on-path observer to the connection on which it was issued.  For example, it
```

cannot include the connection ID or addressing information unless the values are
encrypted.

Unlike the token that is created for a Retry packet, there might be some time
between when the token is created and when the token is subsequently used.
Thus, a token SHOULD have an expiration time, which could be either an explicit
expiration time or an issued timestamp that can be used to dynamically calculate

> the expiration time. A server can store the expiration time or include it in an

Reviewer: ```suggestion
the expiration time.  A server can store the expiration time or include it in an
```

 0                   1                   2                   3
 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+

> |0|1|               Unpredictable Bits (422..)                ...

Reviewer 2: This one remains particularly thorny.  Now there is no way to generate a stateless reset that is a reasonable size.  Yes, this is technically bounded, but now the naive stateless reset is considerably bigger than the packet that triggers it in many more cases.

Author: Isn't there a rule that you can't send a reset that's biger than the originating packet?

Reviewer 3: eternal ping pong rule

Reviewer 3: you can get around sending a reset only once per every 4 received packets on average, but consensus dislike random.

Author: discussed in London; I think we're OK with this

> |0|1|               Unpredictable Bits (422..)                ...

Reviewer: So we're increasing the min size of a stateless reset token?  I'd really rather not, since that increases the chances of that being a DoS vector.  Keeping the ratio of incoming to outgoing bytes ~1:1 seems important for DoS.  As a server operator, I don't want to not respond to a client's PING only packet with a stateless reset just because it's too small, but I also don't want to respond to every small packet I don't have state for.

Reviewer 3: I still think we should allow responding to smaller packets at some random factor rather than not responding at all. This does not mean we should not try to keep resets small though.

Reviewer: #2770 filed as a separate issue for this.

Author: OK, if you agree that this issue is non-blocking I'll proceed.

To entities other than its intended recipient, a stateless reset will appear
to be a packet with a short header.  For the packet to appear as valid, the
Unpredictable Bits field needs to include at least 422 bits of data (or 53
bytes, less the two fixed bits).  This is intended to allow for a Destination
Connection ID of the maximum length permitted, with a minimal packet number, and

> payload.  The Stateless Reset Token corresponds to the minimum expansion of the

Reviewer 2: A bit further down, there is a 40 and 39 that need to be changed to 70 and 69.

  indicates which version of QUIC is in use and determines how the rest of the
  protocol fields are interpreted.

DCIL:

: The byte following the version contains the length in bytes of the destination
  connection ID field that follows it.  This length is encoded as a 8-bit
  unsigned integer.  In QUIC version 1, this value MUST NOT exceed 48 bytes.
  Endpoints that receive a version 1 long header with a value larger than 48
  MUST drop the packet. Servers SHOULD be able to read longer connection IDs
  from other QUIC versions in order to properly form a version negotiation
  packet.

Destination Connection ID:

: The Destination Connection ID field follows the DCIL and is between 0 and 48
  bytes in length. {{negotiating-connection-ids}} describes the use of this
  field in more detail.

SCIL:

> : The byte following the destination Connection ID contains the length in bytes

Reviewer 4: ```suggestion
: The byte following the Destination Connection ID contains the length in bytes
```

>   of the source connection ID field that follows it.  This length is encoded as

Reviewer 4: ```suggestion
  of the Source Connection ID field that follows it.  This length is encoded as
```

and that the Version Negotiation packet was not generated by an off-path
attacker.

As future versions of QUIC may support Connection IDs larger than the version 1

> limit, Version Negotiation Packets will sometimes exceed the Connection ID

Reviewer 4: ```suggestion
limit, Version Negotiation Packets MAY carry Connection IDs that are longer than 48 bytes.
```

Reviewer 2: s/MAY/could - it's not the place of this specification to give permission to break its own rules.

> limit, Version Negotiation Packets could carry Connection IDs that are longer

Reviewer 2: ```suggestion
limit, Version Negotiation packets could carry Connection IDs that are longer
```

> limit, Version Negotiation Packets MAY carry Connection IDs that are longer than

Reviewer 4: ```suggestion
limit, Version Negotiation Packets could carry Connection IDs that are longer than
```
@martinthomson pointed out that this isn't the spec where we need to allow larger CIDs to happen. This merely needs to note that it could happen.

> length limit of 48 bytes in version 1.

Reviewer 4: ```suggestion

```

|                         Version (32)                          |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|    DCIL (8)   |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|               Destination Connection ID (0..384)            ...
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|    SCIL (8)   |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|                 Source Connection ID (0..384)               ...
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|   ODCIL (8)   |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|          Original Destination Connection ID (0..384)        ...
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|                        Retry Token (*)                      ...
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
~~~
{: #retry-format title="Retry Packet"}

A Retry packet (shown in {{retry-format}}) does not contain any protected
fields. The value in the Unused field is selected randomly by the server. In
 addition to the long header, it contains these additional fields:

ODCIL:

> : These bits the length of the Original Destination Connection ID field.  The

Reviewer 5: Typo `These bits the length`

How about mimicking the SCIL field description?
```
ODCIL:

  The byte following the Source Connection ID contains the length in bytes
  of the Original Destination Connection ID field that follows it.  This length
  is encoded as a 8-bit unsigned integer.  In QUIC version 1, this value
  MUST NOT exceed 48 bytes.
```
